DBZ-2582 Updating LSN handling

jgormley6 opened the pull request:
https://issues.redhat.com/browse/DBZ-2582.

This is an example of what would be done if the changes for DBZ-2582 were implemented across the board.
A more general/customizable option should be built instead.

Edit: I updated the PR to make the select something that's configurable. I really do not like any of the names I gave these things (the configuration and the class/variable names), but I am honestly not sure what to call them that is concise yet informative. Suggestions welcome!

jpechane commented:
@jgormley6 Unfortunately the test is broken so there is probably still something wrong.

Overall I like the change and thinking about that I even believe that it should be enabled by default. For that purpose could you please make the new field internal - see `io.debezium.config.Field.createInternal(String)` so we will not document it.

Also provide the default query you use and change the test to verify what happens if the field is set or is not set.

This would give us an opportunity to step back in case there is soemthing disruptive in the change and if not the parameter can be removed in future release.

jgormley6 commented:
@jpechane I'm not entirely sure I understand what is being asked to be done. 

I updated the PR to have that query be the default implementation and added a test that utilizes the default, and another that removes it. Let me know if you wanted something different to be done.

The test error is because I accidentally removed the metronome.pause() and mucked up the polling interval.

gunnarmorling commented on debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnection.java:
```diff
@@ -150,6 +151,26 @@ public Lsn getMaxLsn() throws SQLException {
     }
 
+    public MaxLsnResult getMaxLsnResult(String alternativeMaxTransactionalQuery) throws SQLException {
+        Lsn maxLsn = queryAndMap(GET_MAX_LSN, singleResultMapper(rs -> {
```
Btw. in the Jira statement you originally suggested a single statement to get both LSN values. Can we actually do this, avoiding one round-trip to the DB?

gunnarmorling replied:
Also, should the existing `getMaxLsn()` method be removed, and all callers be updated to use this one?

jgormley6 replied:
I can update to the prepared query.

I'm not sure how we want to handle these sql statements. @jpechane made a valuable suggestion in that we have this feature defaulted, but still removable incase there are issues with it, so I was trying to minimize changing from the current flow. I could update the code to be more of a split between the current way we handle the lsn and the new way and have the new way utilize the query that contains both lsns.

`getMaxLsn()` still has its uses as the stored procs require a valid lsn, which the max transactional way may not be valid after a long period of inactivity. The idea is that one lsn can be utilize for comparison, and the other should be utilized for any queries

gunnarmorling commented on debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnection.java:
```diff
@@ -150,6 +151,26 @@ public Lsn getMaxLsn() throws SQLException {
+            return ret;
+        }, "Maximum LSN query must return exactly one value"));
+
+        if (StringUtils.isEmpty(alternativeMaxTransactionalQuery)) {
```
Seems this would only be the case if someone set the statement explicitly to ""; is this what you're expecting here?

jgormley6 replied:
The original idea for the configuration was that if you put nothing, the current flow wouldn't change, but with this second go around where the query is defaulted, but still overridable, I'm not sure what exactly we want to do.

Maybe we could make this as a sort of feature flag and just have a boolean for determining whether or not we utilize the new lsn handling?

gunnarmorling replied:
I think the question is whether there's a realistic chance for other statements really being used. If not (which I think is the case tbh.) then yes hard-coding the statement and instead having a feature-toggle for enabling/disabling this optimization seems more reasonable. It could remain an internal (undocumented one) as is right now, defaulting to enable this. Then this could be used to disable the optimization should it prove troublesome in production.

gunnarmorling commented on debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/MaxLsnResult.java:
```diff
@@ -0,0 +1,35 @@
+        return maxLsn;
+    }
+
+    public Lsn getMaxTransactionalLsn() {
```
Do we actually need to expose both values to calling code? Wouldn't it suffice to expose only the lower of the two?

jgormley6 replied:
We need both values because the stored proc later used to read the changes will fail if you give it an invalid lsn.
When query the actual tables, we need to always give it the real maximum lsn (and not just the highest one related to changes) otherwise if a server is dormant for a while, you'll run into errors.

I tried clarifying that through the comments, let me know what changes I can make to them as it's not an obvious thing and should be well documented in the code why.

gunnarmorling replied:
I think I got it overall, thanks for elaborating! One thing still unclear: when does an LSN become invalid exactly?

jgormley6 replied:
From my understanding, LSNs are cleaned up and removed from the lsn_time_mapping table. It happens automatically, but I believe this stored procedure also could trigger it manually if you wanted to see an example 
https://docs.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sys-sp-cdc-cleanup-change-table-transact-sql?view=sql-server-ver15

gunnarmorling reviewed (commented):
A few comments/suggestions inline. To be sure, this all is an optimization for the case where there's no traffic in the database at all, correct?

jgormley6 commented:
Yes it is an optimization for when there is no traffic. SqlServer inserts a dummy LSN every 5 minutes or so the process can be verified as "running" but that lsn doesn't map to any changes. 

I'll look into the tests, sorry I haven't been checking and updating this PR too often.

jgormley6 commented:
Not sure what is up with the format check.
`Could not find artifact io.debezium:debezium-core:jar:1.4.0-SNAPSHOT in confluent (https://packages.confluent.io/maven/)` seems like since 1.4.0 hasn't be deployed to that maven repo so it fails.

jpechane commented:
@jgormley6 Thanks for the rpeort. It looks like an edge case captured in https://issues.redhat.com/browse/DBZ-2623

jgormley6 commented:
My rename test is flakey. I've been trying to debug it but unfortunately it only fails in the CI 😭 so it's been difficult to fix.
There seems to be some sort of race condition around CDC being disabled and the querying for changes in the CDC table. 


Going to keep looking at it over the weekend/next week.

jgormley6 commented:
The test has passed consistently 3 times now so I am feeling better about it now. 

If anyone sees issues with the PR (or the test), please let me now

gunnarmorling commented on debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java:
```diff
@@ -442,6 +452,10 @@ public boolean isReadOnlyDatabaseConnection() {
         return readOnlyDatabaseConnection;
     }
 
+    public boolean isSkipLowActivityLSNsEnabled() {
```
Tiny nit: "LSN" -> "Lsn" (this isn't done consistently across the code base yet, but it's the casing we want to converge on eventually for abbreviations in type/method/variable names).

gunnarmorling commented on debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnection.java:
```diff
@@ -153,6 +156,24 @@ public Lsn getMaxLsn() throws SQLException {
     }
 
+    public MaxLsnResult getMaxLsnResult(boolean skipLowActivityLSNsEnabled) throws SQLException {
+        if (skipLowActivityLSNsEnabled) {
```
Same comment as below.

gunnarmorling reviewed (approved):
@jgormley6, thanks for the push; one tiny comment inline, otherwise this is good to go from my PoV for 1.4.0.Alpha1. @jpechane?

jpechane commented:
I agree, this is good to go. Juist please align the names as requested and this can be merged. Thanks a lot for your patience!

gunnarmorling commented:
Sigh. I don't think the test failure is related? @Naros, is this one known/tracked already? @jgormley6, unless you disagree, I' merge this?

jgormley6 commented:
That's the test that's been giving me headaches..... I don't think it's directly related to these changes necessarily, but it's something I'll continue looking into.

gunnarmorling commented:
Rebased and applied; thanks, @jgormley6!

> but it's something I'll continue looking into.

Ok, thanks. Any insights will be welcomed!
